[AC-2828] Add provider portal members page behind FF

amorask-bitwarden opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->

https://bitwarden.atlassian.net/browse/AC-2828

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->

This PR adds a new "Members" page to the Provider Portal behind a FF that pivots between the existing "People" page and the new "Members" page.

Additionally, it migrates several components including `user-add-edit`, `bulk-confirm` and `bulk-remove` to dialogs so that we completely avoid the use of the `modalService` in the new page.

The goal for the new "Members" page is to function very similarly to the "Members" page in Admin Console, but specifically for Provider Users. This includes using the `new-base.people.component`, `bit-table` and moving the bulk functionality into the 3-dot menu in the table header.

### ❓ Open Question
Currently, when you load the new "Members" page with a search query parameter, there's occasionally some flashing that takes place during the filtering. It also happens on the "Members" page in Admin Console. I've narrowed down one possible fix to removing the `debounceTime` from the `searchControl.valueChanges` observable in the `new-base.people.component` here: https://github.com/bitwarden/clients/blob/main/apps/web/src/app/admin-console/common/new-base.people.component.ts#L88-L92

However, I don't want to do so when I'm not sure what it was originally intended for. CC: @eliykat 

## 📸 Screenshots

<!-- Required for any UI changes; delete if not applicable. Use fixed width images for better display. -->

- Adding users
- Editing users
- Re-inviting users
- Confirming users
- Removing users

https://github.com/bitwarden/clients/assets/144709477/31bfcb6d-2696-4da9-a67b-c01102623213

- Bulk re-inviting users

https://github.com/bitwarden/clients/assets/144709477/21637f9b-1931-4f67-b401-aade305d485a

- Bulk confirming users


https://github.com/bitwarden/clients/assets/144709477/e7905301-9b86-4c12-b3af-f6d8e3fc6f29

- Searching users
- Events
- Bulk deleting users


https://github.com/bitwarden/clients/assets/144709477/aaf05d67-c425-4466-b66b-692e9150e251




## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/dialogs/members-dialog.component.ts:
```diff
@@ -0,0 +1,140 @@
+
+      this.dialogRef.close(MembersDialogResultType.Deleted);
+    } catch (error) {
+      this.logService.error(error);
```
Now that the template is using the `bitAction` / `bitSubmit` directives, you don't need this error handling in the method. i.e. delete the try/catch.

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/dialogs/members-dialog.component.ts:
```diff
@@ -0,0 +1,140 @@
+
+      this.dialogRef.close(MembersDialogResultType.Saved);
+    } catch (error) {
+      this.logService.error(error);
```
Same here

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/dialogs/members-dialog.component.html:
I suggest a more descriptive name, it's difficult to know what this is at a glance. `add-edit-member-dialog.component.html`?

eliykat reviewed (changes requested):
Great work! And thanks for doing this, it ticks off a few tickets we had on the AC backlog.

Re: the search control debounce. As discussed, this is to avoid running the filter function on every keystroke, and only run it once the user has stopped (or at least paused) typing.

To avoid the delay while keeping the debounce (which I still prefer), we need to debounce the user input, but not when we programmatically set the value.

We could do this as follows:

```ts
// in the this.route.queryParams subscribe callback
// update the search control value, but don't emit a change event
this.searchControl.setValue(qParams.search, {emitEvent: false});
// update the filter manually (not debounced)
this.dataSource.filter = peopleFilter(qParams.search, null);
```

That would be the same for both subclasses, so it could be moved to the base class.

As an aside, I think the proper flow here is probably search input (debounced) => update queryParams => update filter, effectively storing the search state in the URL. But that requires additional changes that are out of scope here.

amorask-bitwarden commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/people.component.ts:
```diff
@@ -87,6 +90,20 @@ export class PeopleComponent extends BasePeopleComponent<ProviderUserUserDetails
   ngOnInit() {
     // eslint-disable-next-line rxjs-angular/prefer-takeuntil, rxjs/no-async-subscribe
     this.route.parent.params.subscribe(async (params) => {
+      const useProviderPortalMembersPage = await this.configService.getFeatureFlag(
+        FeatureFlag.AC2828_ProviderPortalMembersPage,
+      );
+
+      if (useProviderPortalMembersPage) {
+        const queryParams = await firstValueFrom(this.route.queryParams);
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this.router.navigate(["../members"], {
+          relativeTo: this.route,
+          queryParams,
+        });
+        return;
```
Ah this is what I was looking for, thank you. Out of curiosity though, do you know if there'd be a way to dynamically pivot the `routeOptions.path` and `routeOptions.data.titleId` off the feature flag as well? I ask because part of the requirements for this ticket were renaming the path and page title to "Members" from "People", but I think using `featureFlaggedRoute` would only allow it to be "People".

eliykat replied:
I think there are 2 ways you could achieve this, depending on your preference.

1. submit a standalone PR to change this naming without any of the other changes (i.e. update the path, `titleId`, and component title). I think that would be really minimal and not require any QA or feature flagging. Then you don't need to worry about it in this PR, and use the `featureFlaggedRoute` util as described.

2. keep separate route definitions, and use the [canAccessFeature](https://github.com/bitwarden/clients/blob/main/libs/angular/src/platform/guard/feature-flag.guard.ts) route guard instead. That's a simpler `canActivate` guard that will reject access to the new component unless the flag is on, and optionally redirect you to the old path that you specify. There's no equivalent for the old component, but you could easily create an inverted canActivate function (`cannotAccessFeature`?).

amorask-bitwarden commented on bitwarden_license/bit-web/src/app/admin-console/providers/providers-layout.component.html:
See my question here: https://github.com/bitwarden/clients/pull/9949#discussion_r1666878084

amorask-bitwarden commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/dialogs/bulk-confirm-dialog.component.html:
Yeah this is probably just my lack of knowledge how to accomplish this in a more DRY manner. The AC `BulkConfirmComponent` has dialog params that do not include the `providerId`. In the `bit-web` subclass, it takes the `providerId` as an `@Input()`, which is then passed in through the `setComponentParameters` of the `modalService.openViewRef`. I couldn't determine if something like this was possible with the `dialogService`, or if I just had to extend the base class' dialog parameters to take in a `providerId`, which also felt a bit weird.

eliykat replied:
Hmm yes I understand now.

If you don't want to spend much time on it, then I think you can add `providerId` to the params, which _is_ weird, but I'll fix it up behind you.

My proper recommendation (which is what I'd be doing anyway) is to:
* create a `BaseConfirmComponent` abstract class, which only has abstract method declarations for the methods that the PP component overrides: `isAccepted`, `getPublicKeys`, `getCryptoKey`, `postConfirmRequest`. It does not inject any dialog data.
* have separate components for AC and PP which extend the `BaseConfirmComponent`, implement the abstract methods, and inject their required dialog data.

Same for `BulkRemoveComponent`.

That might mess with your QA scope and feature flagging, so you're welcome to do it if you're keen, but I'm also happy to use the tech debt time to polish that up.

amorask-bitwarden replied:
Took a stab at implementing these [here](https://github.com/bitwarden/clients/pull/9949/commits/ade59407be81b41623e6d97481f08b6a4c5e72da) and [here](https://github.com/bitwarden/clients/pull/9949/commits/f6bef32f0fc16d80528411c39455c71a07667884).

amorask-bitwarden commented:
> Great work! And thanks for doing this, it ticks off a few tickets we had on the AC backlog.
> 
> Re: the search control debounce. As discussed, this is to avoid running the filter function on every keystroke, and only run it once the user has stopped (or at least paused) typing.
> 
> To avoid the delay while keeping the debounce (which I still prefer), we need to debounce the user input, but not when we programmatically set the value.
> 
> We could do this as follows:
> 
> ```ts
> // in the this.route.queryParams subscribe callback
> // update the search control value, but don't emit a change event
> this.searchControl.setValue(qParams.search, {emitEvent: false});
> // update the filter manually (not debounced)
> this.dataSource.filter = peopleFilter(qParams.search, null);
> ```
> 
> That would be the same for both subclasses, so it could be moved to the base class.
> 
> As an aside, I think the proper flow here is probably search input (debounced) => update queryParams => update filter, effectively storing the search state in the URL. But that requires additional changes that are out of scope here.

@eliykat Thank you for the review! I implemented this suggestion and it worked as expected! I also implemented the majority of your feedback, but left a question on the feature-flagged routing.

eliykat reviewed (commented):
Looking good, reviewed all commits to date and resolved/responded to comments.

eliykat commented on apps/web/src/app/admin-console/organizations/members/components/bulk/base.bulk-confirm.component.ts:
Nit: please name this `base-bulk-confirm.component.ts` to fit existing naming patterns. Same with the other components.

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/dialogs/bulk-confirm-dialog.component.html:
These components should go back to sharing the same template as their AC counterparts.

eliykat commented on apps/web/src/app/admin-console/organizations/members/components/bulk/bulk-confirm.component.ts:
Correct me if I'm wrong but I don't think these changes are feature flagged. And because the old PP components subclass this one, it also means the changes are leaking into non-flagged PP code.

I suggest reverting changes to the AC `bulk-confirm.component.ts` and `bulk-remove.component.ts`, so that they're not using the new base class.  That will cause some temporary code duplication, but then you can come back and switch them over to the new base class as part of your flag removal cleanup.

eliykat commented on apps/web/src/app/admin-console/organizations/members/components/bulk/base.bulk-confirm.component.ts:
```diff
@@ -0,0 +1,88 @@
+  protected abstract getPublicKeys(): Promise<
+    ListResponse<OrganizationUserBulkPublicKeyResponse | ProviderUserBulkPublicKeyResponse>
+  >;
+  protected abstract isAccepted(user: BulkUserDetails): boolean;
```
`isAccepted` is no longer used by both subclasses - you use it in the AC component but not the PP component.

I also see you've moved the initialization of `filteredUsers` and `excludedUsers` into the subclasses. I can understand that but I think it makes it unclear if/how these properties must be initialized. The benefit of the old structure was that subclasses didn't have to know about this - they just had to initialize `this.users` and define `this.isAccepted`. I think I prefer that structure, i.e. moving this back to the base class `ngOnInit`:

```ts
    this.excludedUsers = this.users.filter((u) => !this.isAccepted(u));
    this.filteredUsers = this.users.filter((u) => this.isAccepted(u));

    if (this.filteredUsers.length <= 0) {
      this.done = true;
    }
```

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/dialogs/bulk-remove-dialog.component.ts:
```diff
@@ -0,0 +1,52 @@
+
+    this.providerId = dialogParams.providerId;
+    this.users = dialogParams.users;
+    this.showNoMasterPasswordWarning = this.users.some(
+      (u) => u.status > ProviderUserStatusType.Invited && u.hasMasterPassword === false,
```
I think this is from the AC component, not the PP one. It doesn't apply to ProviderUsers.

eliykat reviewed (changes requested):
This is exactly what I was thinking of, just a few changes around implementation details and feature flagging.

amorask-bitwarden commented:
> This is exactly what I was thinking of, just a few changes around implementation details and feature flagging.

Sweet, thanks for catching those - I see what you were saying now. I think I got the remainder of your open comments wrapped up. I also updated https://bitwarden.atlassian.net/browse/AC-2845 to add the task of making the AC `bulk-confirm` and `bulk-remove` components inherit from the new base components.

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/dialogs/bulk-confirm-dialog.component.ts:
```diff
@@ -42,17 +42,6 @@ export class BulkConfirmDialogComponent extends BaseBulkConfirmComponent {
-      this.done = true;
-    }
-
     this.loading = false;
```
`this.loading = false;` should be moved to the bottom of the base class `ngOnInit` method, as we're not actually done loading until that's complete. That seems to match the current implementation in the old AC component.

amorask-bitwarden replied:
Good call - got confused about Angular's order of operations there.

eliykat commented on apps/web/src/app/admin-console/common/base.members.component.ts:
This name also needs to be fixed: `base-members.component.ts`.

amorask-bitwarden replied:
There's several others named this way, but I only renamed the `base.members.component` to stay within scope of this ticket. 
![image](https://github.com/bitwarden/clients/assets/144709477/1fd451f0-faf1-46db-9295-ed8b08d09213)

eliykat replied:
Good point, that's fine.

eliykat commented on bitwarden_license/bit-web/src/app/admin-console/providers/manage/people.component.ts:
```diff
@@ -90,6 +96,20 @@ export class PeopleComponent
   ngOnInit() {
     // eslint-disable-next-line rxjs-angular/prefer-takeuntil, rxjs/no-async-subscribe
     this.route.parent.params.subscribe(async (params) => {
+      const useProviderPortalMembersPage = await this.configService.getFeatureFlag(
+        FeatureFlag.AC2828_ProviderPortalMembersPage,
+      );
+
+      if (useProviderPortalMembersPage) {
+        const queryParams = await firstValueFrom(this.route.queryParams);
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        this.router.navigate(["../members"], {
+          relativeTo: this.route,
+          queryParams,
+        });
+        return;
```
This code can be deleted now that you're using `featureFlaggedRoute`.

amorask-bitwarden replied:
Ah whoops, I forgot to do this in the `members.component` as well.

eliykat reviewed (changes requested):
Really great work here, just a few very minor changes to finish up, then we're good to go.

amorask-bitwarden commented:
@eliykat Thanks for catching those last misses. Think I got the rest of it cleaned up! https://github.com/bitwarden/clients/pull/9949/commits/f8f15f4a80439537f5ef36fa8dde9e77bccbeb94

eliykat reviewed (dismissed):
:shipit:

amorask-bitwarden commented:
@eliykat Sorry had a conflict, can you restamp when you have a sec?
